cli/internal/lockfile/npm_lockfile.go

```diff
@@ -0,0 +1,232 @@
+	Resolved     string                   `json:"resolved,omitempty"`
+	Integrity    string                   `json:"integrity,omitempty"`
+	Bundled      bool                     `json:"bundled,omitempty"`
+	Dev          bool                     `json:"dev,omitempty"`
+	Optional     bool                     `json:"optional,omitempty"`
+	Requires     map[string]string        `json:"requires,omitempty"`
+	Dependencies map[string]NpmDependency `json:"dependencies,omitempty"`
+}
+
+var _ Lockfile = (*NpmLockfile)(nil)
+
+// ResolvePackage Given a workspace, a package it imports and version returns the key, resolved version, and if it was found
+func (l *NpmLockfile) ResolvePackage(workspacePath turbopath.AnchoredUnixPath, name string, version string) (Package, error) {
+	_, ok := l.Packages[workspacePath.ToString()]
+	if !ok {
+		return Package{}, fmt.Errorf("No package found in lockfile for '%s'", workspacePath)
+	}
+
+	// AllDependencies will return a key to avoid choosing the incorrect transitive dep
+	if entry, ok := l.Packages[name]; ok {
+		return Package{
+			Key:     name,
+			Version: entry.Version,
+			Found:   true,
+		}, nil
+	}
+
+	// If we didn't find the entry just using name, then this is an initial call to ResolvePackage
+	// based on information coming from internal packages' package.json
+	// First we check if the workspace uses a nested version of the package
+	nestedPath := fmt.Sprintf("%s/node_modules/%s", workspacePath, name)
+	if entry, ok := l.Packages[nestedPath]; ok {
+		return Package{
+			Key:     nestedPath,
+			Version: entry.Version,
+			Found:   true,
+		}, nil
+	}
+
+	// Next we check for a top level version of the package
+	hoistedPath := fmt.Sprintf("node_modules/%s", name)
+	if entry, ok := l.Packages[hoistedPath]; ok {
+		return Package{
+			Key:     hoistedPath,
+			Version: entry.Version,
+			Found:   true,
+		}, nil
+	}
+
+	return Package{}, nil
+}
+
+// AllDependencies Given a lockfile key return all (dev/optional/peer) dependencies of that package
+func (l *NpmLockfile) AllDependencies(key string) (map[string]string, bool) {
+	entry, ok := l.Packages[key]
+	if !ok {
+		return nil, false
+	}
+	deps := make(map[string]string, len(entry.Dependencies)+len(entry.DevDependencies)+len(entry.PeerDependencies)+len(entry.OptionalDependencies))
+	addDep := func(d map[string]string) {
+		for name := range d {
+			for _, possibleKey := range possibleNpmDeps(key, name) {
+				if entry, ok := l.Packages[possibleKey]; ok {
+					deps[possibleKey] = entry.Version
+					break
+				}
+			}
+		}
+	}
+
+	addDep(entry.Dependencies)
+	addDep(entry.DevDependencies)
+	addDep(entry.OptionalDependencies)
+	addDep(entry.PeerDependencies)
+
+	return deps, true
+}
+
+// Subgraph Given a list of lockfile keys returns a Lockfile based off the original one that only contains the packages given
+func (l *NpmLockfile) Subgraph(workspacePackages []turbopath.AnchoredSystemPath, packages []string) (Lockfile, error) {
```

Comment:
if the pattern isn't already set, maybe `PrunedLockfile` instead of `Subgraph`? (rather than giving a new name to the same concept)

Reply:
`Subgraph` was the name suggested by @nathanhammond back in #1789